# ChemTONIC: A Brief Guideline and Examples

![Logo](https://raw.githubusercontent.com/mldlproject/chemtonic/e218517eb2b5f73553035badad7e32f0f37bd291/chemtonic.svg)

### PART 1: FUNCTIONS IN SUBSTAGE MODULES

ChemTONIC 0.0.1 was released with the first module called **curation**. Module **curation** contains four submodules, including
- `validation`
- `cleaning`
- `normalization`
- `refinement`
- `utils`

This brief guideline is designed to help novice users to perform data curation in the fastest way. For the details of all provided functions, please read "[ChemTONIC: A Complete Guideline and Examples](https://github.com/mldlproject/chemtonic/blob/main/CompleteGuideline.ipynb)"

### Import libraries

In [ ]:
from IPython.display import Image
from IPython.core.display import HTML 

In [6]:
from chemtonic.curation.validation import validateComplete
from chemtonic.curation.cleaning import cleanComplete
from chemtonic.curation.normalization import normalizeComplete
from chemtonic.curation.refinement import refineComplete
import pandas as pd

- validateComplete(): Perform validation stages including all substages
- cleanComplete(): Perform cleaning stages including all substages
- normalizeComplete(): Perform normalization stages including all substages
- refineComplete(): Perform complete data curation processincluding all stages

![Pipeline](https://raw.githubusercontent.com/mldlproject/chemtonic/0fdf3801a315d23cee264535b5bc13bea1b83572/images/pipelinesmall.svg)

### Load example dataset

In [ ]:
smiles = pd.read_csv("./data/example.csv")['SMILES']

### 1. Validation

Module **validation** contains four functions: *rmMixtures()*, *rmInorganics()*, *rmOrganometallics()*, and *validateComplete()*. ***validateComplete()*** performs complete validation substage, including *rmMixtures()*,  *rmInorganics()*, and *rmOrganometallics()*

```python
validateComplete(compounds, getInvalidStruct=False, removeDuplicates=False, getDuplicatedIdx=False, exportCSV=False, outputPath=None, printlogs=True)
```


Function arguments:
- ``compounds``: Input compounds. Can be in the form of `str`, `list`, `pandas.core.series.Series`, `pandas.core.frame.DataFrame` 
- ``getInvalidStruct``: Boolean. Default `False`. Set `True` to get unvalidated SMILES (if any)
- ``removeDuplicates``: Boolean. Default `False`. Set `True` to remove duplicated SMILES (if any)
- ``getDuplicatedIdx``: Boolean. Default `False`. Set `True` to get index of duplicated SMILES (if any). Can be used when `removeDuplicates=True` only.
- ``exportCSV``: Boolean. Default `False`. Set `True` to get csv file
- ``outputPath``: Directory. Must be filled when `exportCSV=True` to get csv file
- ``printlogs``: Boolean. Default `True`. Print logs and summary during the process

#### 1.1. Obtaining lists of validated SMILES(s) (duplicated data may exist)

In [9]:
validateComplete(smiles, removeDuplicates=False)

Succeeded to validate 500/500 structures 

499/500 structures are non-mixtures
1/500 structure(s) is/are mixture(s) and was/were removed 

500/500 structures are organics 

500/500 structures are NOT organometallics 

SUMMARY:
499/500 structures were successfully validated
1/500 structure(s) were/was unsuccessfully validated and need to be rechecked
-----------------------------------------------------------------------------
set 'getInvalidStruct=True' to get the list of all unvalidated structures 



SMILES
0              Cc1cc2c(cc1O)C1(C)CCC(O)C(C)(C)C1=CC2=O
1    CC1(C)CCC2(C(=O)O)CCC3(C)C(=CCC4C5(C)CCC(OC6OC...
2                        COC1CC(=O)CCC1(O)CC=Cc1ccccc1
3      C=CC(C)=CCC1C(=C)CCC2C(C)(C)C(O)C(OC(C)=O)CC12C
4                       COc1ccccc1C=CCC1(O)C=CC(=O)CC1
..                                                 ...
494                 COc1ccc(C=CCc2ccc(O)c(OC)c2OC)cc1O
495                C=CCc1ccc(O)c(-c2ccc(O)c(CC=C)c2)c1
496  COc1cc(C2Oc3cc(C4Oc5cc(O)cc(O)c5C(=O)C4O)ccc3O...
497        COc1cc2c(cc1OC)C1C(C)c3ccc(OC)c(OC)c3CN1CC2
498                     COC(c1ccccc1)C(O)C1CC=CC(=O)O1

[499 rows x 1 columns]

#### 1.2. Obtaining lists of validated SMILES(s) (duplicated data removed)

In [10]:
validateComplete(smiles, removeDuplicates=True)

Succeeded to validate 500/500 structures 

499/500 structures are non-mixtures
1/500 structure(s) is/are mixture(s) and was/were removed 

500/500 structures are organics 

500/500 structures are NOT organometallics 

SUMMARY:
499/500 structures were successfully validated
1/500 structure(s) were/was unsuccessfully validated and need to be rechecked
-----------------------------------------------------------------------------
set 'getInvalidStruct=True' to get the list of all unvalidated structures 

There are 381 unique structures filtered from 499 initial validated structures
-----------------------------------------------------------------------------
To get detailed information, please follow steps below:
(1) Rerun validateComplete() with setting 'removeDuplicates=False' to get the list of all validated structures
(2) Run ultils.molDeduplicate() with setting 'getDuplicates=True'to get the list of duplicated structures 

--OR--
Rerun validateComplete() with setting 'getDuplicates=Tr

SMILES
0              Cc1cc2c(cc1O)C1(C)CCC(O)C(C)(C)C1=CC2=O
1    CC1(C)CCC2(C(=O)O)CCC3(C)C(=CCC4C5(C)CCC(OC6OC...
2                        COC1CC(=O)CCC1(O)CC=Cc1ccccc1
3      C=CC(C)=CCC1C(=C)CCC2C(C)(C)C(O)C(OC(C)=O)CC12C
4                       COc1ccccc1C=CCC1(O)C=CC(=O)CC1
..                                                 ...
376  CC(=O)C1c2oc3c4c(cc(O)c3c(=O)c2CC2C1C(=O)OC2(C...
377  CC(C)=CCCC(C)=CCc1c(O)ccc2cc(-c3cc(O)cc(O)c3)oc12
378                 COc1ccc(C=CCc2ccc(O)c(OC)c2OC)cc1O
379        COc1cc2c(cc1OC)C1C(C)c3ccc(OC)c(OC)c3CN1CC2
380                     COC(c1ccccc1)C(O)C1CC=CC(=O)O1

[381 rows x 1 columns]

#### 1.3. Obtaining lists of unvalidated SMILES(s) (if any)

In [11]:
validateComplete(smiles, getInvalidStruct=True)

Succeeded to validate 500/500 structures 

499/500 structures are non-mixtures
1/500 structure(s) is/are mixture(s) and was/were removed 

500/500 structures are organics 

500/500 structures are NOT organometallics 

SUMMARY:
499/500 structures were successfully validated
1/500 structure(s) were/was unsuccessfully validated and need to be rechecked
-----------------------------------------------------------------------------


SMILES errorTag  fromFunction  \
0  CCC1(O)CC2CN(CCc3c([nH]c4ccccc34)C(C(=O)OC)(c3...  Mixture  rmMixtures()   

   idx  
0  277

#### 1.4. Obtaining lists of duplicated validated SMILES(s) (if any)

In [18]:
validateComplete(smiles, removeDuplicates=True, getDuplicatedIdx=True)

Succeeded to validate 500/500 structures 

499/500 structures are non-mixtures
1/500 structure(s) is/are mixture(s) and was/were removed 

500/500 structures are organics 

500/500 structures are NOT organometallics 

SUMMARY:
499/500 structures were successfully validated
1/500 structure(s) were/was unsuccessfully validated and need to be rechecked
-----------------------------------------------------------------------------
set 'getInvalidStruct=True' to get the list of all unvalidated structures 

There are 381 unique structures filtered from 499 initial validated structures
-----------------------------------------------------------------------------
To get detailed information, please follow steps below:
(1) Rerun validateComplete() with setting 'removeDuplicates=False' to get the list of all validated structures
(2) Run ultils.molDeduplicate() with setting 'getDuplicates=True'to get the list of duplicated structures 

--OR--
Rerun validateComplete() with setting 'getDuplicates=Tr

idx       matchedIdx        fromFunction
0      0  [0, 9, 69, 278]  validateComplete()
1      3         [3, 467]  validateComplete()
2      7         [7, 215]  validateComplete()
3      9  [0, 9, 69, 278]  validateComplete()
4     11        [11, 418]  validateComplete()
..   ...              ...                 ...
196  487       [464, 487]  validateComplete()
197  492       [200, 492]  validateComplete()
198  493       [445, 493]  validateComplete()
199  495       [213, 495]  validateComplete()
200  496       [472, 496]  validateComplete()

[201 rows x 3 columns]

#### 1.5. Export csv files

To create csv file, run the commands below with `outputPath` defined.

In [ ]:
#ValOutput= '../validation/'
# validateComplete(smiles, removeDuplicates=False, exportCSV=True, outputPath=ValOutput)
# validateComplete(smiles, removeDuplicates=True,  exportCSV=True, outputPath=ValOutput)
# validateComplete(smiles, getInvalidStruct=True,  exportCSV=True, outputPath=ValOutput)
# validateComplete(smiles, removeDuplicates=True,  getDuplicatedIdx=True, exportCSV=True, outputPath=ValOutput)

---

### 2. Cleaning

Module **cleaning** contains two functions: *clSalts()*, *clCharges()*, and *cleanComplete()*. ***cleanComplete()*** performs complete cleaning substage, including *clSalts()* and *clCharges()*.

**Notice**: Cleaning substage requires all SMILESs to be verified. In case of unverified SMILES(s) available, cleanComplete() will remove unvalidated SMILES(s). cleanComplete() does not support removing mixtures, inorganics, or organometallics. 

```python
cleanComplete(compounds, getUncleanedStruct=False, deSalt=False, neutralize=False, removeDuplicates=False, getDuplicatedIdx=False, exportCSV=False, outputPath=None, printlogs=True)
```


Function arguments:
- ``compounds``: Input compounds. Can be in the form of `str`, `list`, `pandas.core.series.Series`, `pandas.core.frame.DataFrame` 
- ``getUncleanedStruct``: boolean. Default `False`. Set `True` to get uncleaned SMILES (if any)
- ``deSalt``: boolean. Default `False`. Set `True` to get the SMILES desalted (if any). Default `False` removes SMILES(s) of salt(s).
- ``neutralize``: boolean. Default `False`. Set `True` to get the SMILES neutralized (if any)
- ``removeDuplicates``: boolean. Default `False`. Set `True` to remove duplicated SMILES (if any)
- ``getDuplicatedIdx``: boolean. Default `False`. Set `True` to get index of duplicated SMILES (if any). Can be used when `removeDuplicates=True` only.
- ``exportCSV``: boolean. Default `False`. Set `True` to get csv file
- ``outputPath``: Directory. Must be filled when `exportCSV=True` to get csv file
- ``printlogs``: Print logs, summary during the process

#### 2.1. Obtaining lists of cleaned SMILES(s) (duplicated data may exist)

In [15]:
cleanComplete(smiles, removeDuplicates=False)

Succeeded to validate 500/500 structures 

497/500 structures are NOT salts
3/500 structure(s) is/are salt(s) and was/were removed 

496/500 structures are NOT charges
4/500 structure(s) is/are charge(s) BUT was/were NOT neutralized 

SUMMARY:
500/500 structures were successfully verfied
497/500 structures were successfully cleaned
3/500 structure(s) was/were unsuccessfully cleaned and need to be rechecked
-------------------------------------------------------
set 'getUncleanedStruct=True' to get the list of all uncleaned structures. Neutralized charged structures will be included (if any) 



SMILES
0              Cc1cc2c(cc1O)C1(C)CCC(O)C(C)(C)C1=CC2=O
1    CC1(C)CCC2(C(=O)O)CCC3(C)C(=CCC4C5(C)CCC(OC6OC...
2                        COC1CC(=O)CCC1(O)CC=Cc1ccccc1
3      C=CC(C)=CCC1C(=C)CCC2C(C)(C)C(O)C(OC(C)=O)CC12C
4                       COc1ccccc1C=CCC1(O)C=CC(=O)CC1
..                                                 ...
492                 COc1ccc(C=CCc2ccc(O)c(OC)c2OC)cc1O
493                C=CCc1ccc(O)c(-c2ccc(O)c(CC=C)c2)c1
494  COc1cc(C2Oc3cc(C4Oc5cc(O)cc(O)c5C(=O)C4O)ccc3O...
495        COc1cc2c(cc1OC)C1C(C)c3ccc(OC)c(OC)c3CN1CC2
496                     COC(c1ccccc1)C(O)C1CC=CC(=O)O1

[497 rows x 1 columns]

#### 2.2. Obtaining lists of cleaned SMILES(s) (duplicated data removed)

In [16]:
cleanComplete(smiles, removeDuplicates=True)

Succeeded to validate 500/500 structures 

497/500 structures are NOT salts
3/500 structure(s) is/are salt(s) and was/were removed 

496/500 structures are NOT charges
4/500 structure(s) is/are charge(s) BUT was/were NOT neutralized 

SUMMARY:
500/500 structures were successfully verfied
497/500 structures were successfully cleaned
3/500 structure(s) was/were unsuccessfully cleaned and need to be rechecked
-------------------------------------------------------
set 'getUncleanedStruct=True' to get the list of all uncleaned structures. Neutralized charged structures will be included (if any) 

There are 380 unique structures filtered from 497 initial cleaned structures
-----------------------------------------------------------------------------
To get detailed information, please follow steps below:
(1) Rerun cleanComplete() with setting 'removeDuplicates=False' to get the list of all validated structures
(2) Run ultils.molDeduplicate() with setting 'getDuplicates=True'to get the list 

SMILES
0              Cc1cc2c(cc1O)C1(C)CCC(O)C(C)(C)C1=CC2=O
1    CC1(C)CCC2(C(=O)O)CCC3(C)C(=CCC4C5(C)CCC(OC6OC...
2                        COC1CC(=O)CCC1(O)CC=Cc1ccccc1
3      C=CC(C)=CCC1C(=C)CCC2C(C)(C)C(O)C(OC(C)=O)CC12C
4                       COc1ccccc1C=CCC1(O)C=CC(=O)CC1
..                                                 ...
375  CC(=O)C1c2oc3c4c(cc(O)c3c(=O)c2CC2C1C(=O)OC2(C...
376  CC(C)=CCCC(C)=CCc1c(O)ccc2cc(-c3cc(O)cc(O)c3)oc12
377                 COc1ccc(C=CCc2ccc(O)c(OC)c2OC)cc1O
378        COc1cc2c(cc1OC)C1C(C)c3ccc(OC)c(OC)c3CN1CC2
379                     COC(c1ccccc1)C(O)C1CC=CC(=O)O1

[380 rows x 1 columns]

#### 2.3. Obtaining lists of uncleaned SMILES(s) (if any)

In [17]:
cleanComplete(smiles, getUncleanedStruct=True)

Succeeded to validate 500/500 structures 

497/500 structures are NOT salts
3/500 structure(s) is/are salt(s) and was/were removed 

496/500 structures are NOT charges
4/500 structure(s) is/are charge(s) BUT was/were NOT neutralized 

SUMMARY:
500/500 structures were successfully verfied
497/500 structures were successfully cleaned
3/500 structure(s) was/were unsuccessfully cleaned and need to be rechecked
-------------------------------------------------------


SMILES errorTag fromFunction  \
0  CCC1(O)CC2CN(CCc3c([nH]c4ccccc34)C(C(=O)OC)(c3...     Salt    clSalts()   
1                                 CCCC(=O)[O-].[Na+]     Salt    clSalts()   
2                                 CCCC(=O)[O-].[Na+]     Salt    clSalts()   
3            COc1ccc(C2CC(=O)c3c([O-])cc(O)cc3O2)cc1   Charge  clCharges()   
4  O=C(O)C1=CC(=CC=[N+]2c3cc(O)c(OC4OC(CO)C(O)C(O...   Charge  clCharges()   
5                                 CCCC(=O)[O-].[Na+]   Charge  clCharges()   
6                                 CCCC(=O)[O-].[Na+]   Charge  clCharges()   

   idx  
0  277  
1  344  
2  379  
3  302  
4  329  
5  344  
6  379

#### 2.4. Obtaining lists of duplicated cleaned SMILES(s) (if any)

In [19]:
cleanComplete(smiles, removeDuplicates=True, getDuplicatedIdx=True)

Succeeded to validate 500/500 structures 

497/500 structures are NOT salts
3/500 structure(s) is/are salt(s) and was/were removed 

496/500 structures are NOT charges
4/500 structure(s) is/are charge(s) BUT was/were NOT neutralized 

SUMMARY:
500/500 structures were successfully verfied
497/500 structures were successfully cleaned
3/500 structure(s) was/were unsuccessfully cleaned and need to be rechecked
-------------------------------------------------------
set 'getUncleanedStruct=True' to get the list of all uncleaned structures. Neutralized charged structures will be included (if any) 

There are 380 unique structures filtered from 497 initial cleaned structures
-----------------------------------------------------------------------------
To get detailed information, please follow steps below:
(1) Rerun cleanComplete() with setting 'removeDuplicates=False' to get the list of all validated structures
(2) Run ultils.molDeduplicate() with setting 'getDuplicates=True'to get the list 

idx       matchedIdx     fromFunction
0      0  [0, 9, 69, 278]  cleanComplete()
1      3         [3, 465]  cleanComplete()
2      7         [7, 215]  cleanComplete()
3      9  [0, 9, 69, 278]  cleanComplete()
4     11        [11, 416]  cleanComplete()
..   ...              ...              ...
194  485       [462, 485]  cleanComplete()
195  490       [200, 490]  cleanComplete()
196  491       [443, 491]  cleanComplete()
197  493       [213, 493]  cleanComplete()
198  494       [470, 494]  cleanComplete()

[199 rows x 3 columns]

#### 2.5. Export csv files

To create csv file, run the commands below with `outputPath` defined.

In [1]:
#CleanOutput= '../cleaning/'
# cleanComplete(smiles, removeDuplicates=False,  exportCSV=True, outputPath=CleanOutput)
# cleanComplete(smiles, removeDuplicates=True,   exportCSV=True, outputPath=CleanOutput)
# cleanComplete(smiles, getUncleanedStruct=True, exportCSV=True, outputPath=CleanOutput)
# cleanComplete(smiles, removeDuplicates=True,   getDuplicatedIdx=True, exportCSV=True, outputPath=CleanOutput)

---

### 3. Normalization

Module `normalization` contains three functions: `normTautomers()`, `normStereoisomers()`, and `normalizeComplete()`.
- `normTautomers()`: remove/detautomerize SMILES of tautomer
- `normStereoisomers()`: destereoisomerize SMILES of stereoisomer
- `normalizeComplete()`: perform complete normalization substage which includes these two steps above

```python
normalizeComplete(compounds, getUnnormalizedStruct=False, deTautomerize=True, deSterioisomerize=True, removeDuplicates=False, getDuplicatedIdx=False, exportCSV=False, outputPath=None, printlogs=True)
```

Function arguments:
- ``compounds``: Input compounds. Can be in the form of `str`, `list`, `pandas.core.series.Series`, `pandas.core.frame.DataFrame` 
- ``getUnnormalizedStruct``:boolean. Default `False`. Set `True` to get unnormalized SMILES (if any)
- ``deTautomerize``: boolean. Default `True`. Set `False` to remove tautomer SMILES (if any)
- ``deSterioisomerize``: boolean. Default `True`. Set `False` to not to desterioisomerize SMILES (if any)
- ``removeDuplicates``: boolean. Default `False`. Set `True` to remove duplicated SMILES (if any)
- ``getDuplicatedIdx``: boolean. Default `False`. Set `True` to get index of duplicated SMILES (if any). Can be used when `removeDuplicates=True` only.
- ``exportCSV``: boolean. Default `False`. Set `True` to get csv file
- ``outputPath``: Directory. Must be filled when `exportCSV=True` to get csv file
- ``printlogs``: Print logs, summary during the process

#### 3.1. Obtaining lists of normalized SMILES(s) (duplicated data may exist)

In [20]:
normalizeComplete(smiles, removeDuplicates=False) 

Succeeded to validate 500/500 structures 

498/500 structures are NOT stereoisomers
2/500 structure(s) is/are stereoisomer(s) BUT was/were destereoisomerized 

437/500 structures are NOT tautomers
63/500 structure(s) is/are tautomer(s) BUT was/were detautomerized 

SUMMARY:
500/500 structure were successfully verfied
500/500 structures were successfully normalized
set 'getUnnormalizedStruct=True' to get the list of all unnormalized structures. 



SMILES
0              Cc1cc2c(cc1O)C1(C)CCC(O)C(C)(C)C1=CC2=O
1    CC1(C)CCC2(C(=O)O)CCC3(C)C(=CCC4C5(C)CCC(OC6OC...
2                        COC1CC(=O)CCC1(O)CC=Cc1ccccc1
3      C=CC(C)=CCC1C(=C)CCC2C(C)(C)C(O)C(OC(C)=O)CC12C
4                       COc1ccccc1C=CCC1(O)C=CC(=O)CC1
..                                                 ...
495                 COc1ccc(C=CCc2ccc(O)c(OC)c2OC)cc1O
496                C=CCc1ccc(O)c(-c2ccc(O)c(CC=C)c2)c1
497  COc1cc(C2Oc3cc(C4Oc5cc(O)cc(O)c5C(=O)C4O)ccc3O...
498        COc1cc2c(cc1OC)C1C(C)c3ccc(OC)c(OC)c3CN1CC2
499                     COC(c1ccccc1)C(O)C1C=CCC(=O)O1

[500 rows x 1 columns]

#### 3.2. Obtaining lists of normalized SMILES(s) (duplicated data removed)

In [21]:
normalizeComplete(smiles, removeDuplicates=True) 

Succeeded to validate 500/500 structures 

498/500 structures are NOT stereoisomers
2/500 structure(s) is/are stereoisomer(s) BUT was/were destereoisomerized 

437/500 structures are NOT tautomers
63/500 structure(s) is/are tautomer(s) BUT was/were detautomerized 

SUMMARY:
500/500 structure were successfully verfied
500/500 structures were successfully normalized
set 'getUnnormalizedStruct=True' to get the list of all unnormalized structures. 

There are 381 unique structures filtered from 500 initial normalized structures
To get detailed information, please follow steps below:
(1) Rerun normalizeComplete() with setting 'removeDuplicates=False' to get the list of all normalized structures
(2) Run ultils.molDeduplicate() with setting 'getDuplicates=True'to get the list of duplicated structures 

--OR--
Rerun normalizeComplete() with setting 'getDuplicates=True', 'exportCSV'=True, and 'outputPath=<Directory>' to export a csv file  containing the list of duplicated structures 

--OR--
Ru

SMILES
0              Cc1cc2c(cc1O)C1(C)CCC(O)C(C)(C)C1=CC2=O
1    CC1(C)CCC2(C(=O)O)CCC3(C)C(=CCC4C5(C)CCC(OC6OC...
2                        COC1CC(=O)CCC1(O)CC=Cc1ccccc1
3      C=CC(C)=CCC1C(=C)CCC2C(C)(C)C(O)C(OC(C)=O)CC12C
4                       COc1ccccc1C=CCC1(O)C=CC(=O)CC1
..                                                 ...
376  CC(O)c1c2c(cc3c(=O)c4c(O)cc5c(c4oc13)C=CC(C)(C...
377  CC(C)=CCCC(C)=CCc1c(O)ccc2cc(-c3cc(O)cc(O)c3)oc12
378                 COc1ccc(C=CCc2ccc(O)c(OC)c2OC)cc1O
379        COc1cc2c(cc1OC)C1C(C)c3ccc(OC)c(OC)c3CN1CC2
380                     COC(c1ccccc1)C(O)C1C=CCC(=O)O1

[381 rows x 1 columns]

#### 3.3. Obtaining lists of unnormalized SMILES(s) (if any)

In [22]:
normalizeComplete(smiles, getUnnormalizedStruct=True) 

Succeeded to validate 500/500 structures 

498/500 structures are NOT stereoisomers
2/500 structure(s) is/are stereoisomer(s) BUT was/were destereoisomerized 

437/500 structures are NOT tautomers
63/500 structure(s) is/are tautomer(s) BUT was/were detautomerized 

SUMMARY:
500/500 structure were successfully verfied
500/500 structures were successfully normalized


SMILES      errorTag  \
0   CC1CN2CC3CCC(=O)C4=C5C(=C(O)C1CC2C53C)C1=C4CCO...      Tautomer   
1         O=C1NCc2c1c1c3ccccc3[nH]c1c1[nH]c3ccccc3c21      Tautomer   
2   COC(=O)C1=C(CCO)C2=C3C1=C(O)C1(O)CC4N(CC1C)CC(...      Tautomer   
3   CC(C)=CCCC1(C)C(CC=C(C)C)CC2(CC=C(C)C)C(=O)C1(...      Tautomer   
4      CC(C)C1=CC2=CC(=O)C3C(C)(C)CCCC3(C)C2=C(O)C1=O      Tautomer   
..                                                ...           ...   
60  CC(=O)CC1C(=O)C(CO)=CCC(C)(CCC(O)C(C)(C)O)C1C=...      Tautomer   
61  CC(=O)C1c2oc3c4c(cc(O)c3c(=O)c2CC2C1C(=O)OC2(C...      Tautomer   
62                     COC(c1ccccc1)C(O)C1CC=CC(=O)O1      Tautomer   
63                F[C@@]12C[C@]1(Cl)C[C@H](/C=C/Br)O2  Stereoisomer   
64                F[C@]12C[C@]1(Cl)C[C@@H](/C=C\Br)O2  Stereoisomer   

           fromFunction  idx  
0       normTautomers()    5  
1       normTautomers()    6  
2       normTautomers()   27  
3       normTautomers()   30  
4       normTautomers()   45  
..                  ...  ...  
60      normTautomers()  486  
61      normTautomers()  491  
62      normTautomers()  499  
63  normStereoisomers()   89  
64  normStereoisomers()   98  

[65 rows x 4 columns]

#### 3.4. Obtaining lists of duplicated normalized SMILES(s) (if any)

In [23]:
normalizeComplete(smiles, removeDuplicates=True, getDuplicatedIdx=True) 

Succeeded to validate 500/500 structures 

498/500 structures are NOT stereoisomers
2/500 structure(s) is/are stereoisomer(s) BUT was/were destereoisomerized 

437/500 structures are NOT tautomers
63/500 structure(s) is/are tautomer(s) BUT was/were detautomerized 

SUMMARY:
500/500 structure were successfully verfied
500/500 structures were successfully normalized
set 'getUnnormalizedStruct=True' to get the list of all unnormalized structures. 

There are 381 unique structures filtered from 500 initial normalized structures
To get detailed information, please follow steps below:
(1) Rerun normalizeComplete() with setting 'removeDuplicates=False' to get the list of all normalized structures
(2) Run ultils.molDeduplicate() with setting 'getDuplicates=True'to get the list of duplicated structures 

--OR--
Rerun normalizeComplete() with setting 'getDuplicates=True', 'exportCSV'=True, and 'outputPath=<Directory>' to export a csv file  containing the list of duplicated structures 

--OR--
Ru

idx       matchedIdx         fromFunction
0      0  [0, 9, 69, 279]  normalizeComplete()
1      3         [3, 468]  normalizeComplete()
2      7         [7, 215]  normalizeComplete()
3      9  [0, 9, 69, 279]  normalizeComplete()
4     11        [11, 419]  normalizeComplete()
..   ...              ...                  ...
198  488       [465, 488]  normalizeComplete()
199  493       [200, 493]  normalizeComplete()
200  494       [446, 494]  normalizeComplete()
201  496       [213, 496]  normalizeComplete()
202  497       [473, 497]  normalizeComplete()

[203 rows x 3 columns]

#### 3.5. Export csv files

In [2]:
#NormOutput= '../normalization/'
# normalizeComplete(smiles, removeDuplicates=False,     exportCSV=True, outputPath=NormOutput)
# normalizeComplete(smiles, removeDuplicates=True,      exportCSV=True, outputPath=NormOutput)
# normalizeComplete(smiles, getUnnormalizedStruct=True, exportCSV=True, outputPath=NormOutput)
# normalizeComplete(smiles, removeDuplicates=True,      getDuplicatedIdx=True, exportCSV=True, outputPath=NormOutput)

---

### 4. Refinement

Module **refinment** contains only one function: refineComplete().
- *refineComplete()*: perform complete data curation process which includes these three substages above

```python
refineComplete(compounds, getUnrefinedStruct=False, deSalt=False, neutralize=False, deTautomerize=True, deSterioisomerize=True, removeDuplicates=True, getDuplicatedIdx=False, exportCSV=False, outputPath=None, printlogs=True):

```

Function arguments:
- ``compounds``: Input compounds. Can be in the form of `str`, `list`, `pandas.core.series.Series`, `pandas.core.frame.DataFrame` 
- ``getUnrefinedStruct``:boolean. Default `False`. Set `True` to get unnormalized SMILES (if any)
-``deSalt``:
-``neutralize``:
- ``deTautomerize``: boolean. Default `True`. Set `False` to remove tautomer SMILES (if any)
- ``deSterioisomerize``: boolean. Default `True`. Set `False` to not to desterioisomerize SMILES (if any)
- ``removeDuplicates``: boolean. Default `True`. Set `False` to keep duplicated SMILES (if any)
- ``getDuplicatedIdx``: boolean. Default `False`. Set `True` to get index of duplicated SMILES (if any). Can be used when `removeDuplicates=True` only.
- ``exportCSV``: boolean. Default `False`. Set `True` to get csv file
- ``outputPath``: Directory. Must be filled when `exportCSV=True` to get csv file
- ``printlogs``: Print logs, summary during the process

#### 4.1. Obtaining lists of normalized SMILES(s) (duplicated data may exist)

In [25]:
refineComplete(smiles, removeDuplicates=False)

VALIDATION
-----------------------------------------------------------------------------
499/500 structures were successfully validated
1/500 structure(s) was/were unsuccessfully validated and need to be rechecked
-----------------------------------------------------------------------------
201/499 validated structures have at least one duplicates
CLEANING
-----------------------------------------------------------------------------
497/499 structures were successfully cleaned
2/499 structure(s) was/were unsuccessfully cleaned and need to be rechecked
-----------------------------------------------------------------------------
No duplicate was found (in 497 cleaned structures)
NORMALIZATION
-----------------------------------------------------------------------------
497/497 structures were successfully normalized
-----------------------------------------------------------------------------
2/497 normalized structures have at least one duplicates
REFINEMENT SUMMARY
-------------------

SMILES
0              Cc1cc2c(cc1O)C1(C)CCC(O)C(C)(C)C1=CC2=O
1    CC1(C)CCC2(C(=O)O)CCC3(C)C(=CCC4C5(C)CCC(OC6OC...
2                        COC1CC(=O)CCC1(O)CC=Cc1ccccc1
3      C=CC(C)=CCC1C(=C)CCC2C(C)(C)C(O)C(OC(C)=O)CC12C
4                       COc1ccccc1C=CCC1(O)C=CC(=O)CC1
..                                                 ...
492                 COc1ccc(C=CCc2ccc(O)c(OC)c2OC)cc1O
493                C=CCc1ccc(O)c(-c2ccc(O)c(CC=C)c2)c1
494  COc1cc(C2Oc3cc(C4Oc5cc(O)cc(O)c5C(=O)C4O)ccc3O...
495        COc1cc2c(cc1OC)C1C(C)c3ccc(OC)c(OC)c3CN1CC2
496                     COC(c1ccccc1)C(O)C1C=CCC(=O)O1

[497 rows x 1 columns]

#### 4.2. Obtaining lists of normalized SMILES(s) (duplicated data removed)

In [26]:
refineComplete(smiles, removeDuplicates=True)

VALIDATION
-----------------------------------------------------------------------------
499/500 structures were successfully validated
1/500 structure(s) was/were unsuccessfully validated and need to be rechecked
-----------------------------------------------------------------------------
There are 381 unique structures filtered from 499 initial validated structures
CLEANING
-----------------------------------------------------------------------------
380/381 structures were successfully cleaned
1/381 structure(s) was/were unsuccessfully cleaned and need to be rechecked
-----------------------------------------------------------------------------
No duplicate was found (in 380 cleaned structures)
NORMALIZATION
-----------------------------------------------------------------------------
379/380 structures were successfully normalized
1/380 structure(s) were unsuccessfully normalized and need to be rechecked
-----------------------------------------------------------------------------

SMILES
0              Cc1cc2c(cc1O)C1(C)CCC(O)C(C)(C)C1=CC2=O
1    CC1(C)CCC2(C(=O)O)CCC3(C)C(=CCC4C5(C)CCC(OC6OC...
2                        COC1CC(=O)CCC1(O)CC=Cc1ccccc1
3      C=CC(C)=CCC1C(=C)CCC2C(C)(C)C(O)C(OC(C)=O)CC12C
4                       COc1ccccc1C=CCC1(O)C=CC(=O)CC1
..                                                 ...
374  CC(O)c1c2c(cc3c(=O)c4c(O)cc5c(c4oc13)C=CC(C)(C...
375  CC(C)=CCCC(C)=CCc1c(O)ccc2cc(-c3cc(O)cc(O)c3)oc12
376                 COc1ccc(C=CCc2ccc(O)c(OC)c2OC)cc1O
377        COc1cc2c(cc1OC)C1C(C)c3ccc(OC)c(OC)c3CN1CC2
378                     COC(c1ccccc1)C(O)C1C=CCC(=O)O1

[379 rows x 1 columns]

#### 4.3 Obtaining lists of unrefined SMILES(s) (if any)

In [28]:
refineComplete(smiles, removeDuplicates=False, getUnrefinedStruct=True)

VALIDATION
-----------------------------------------------------------------------------
499/500 structures were successfully validated
1/500 structure(s) was/were unsuccessfully validated and need to be rechecked
-----------------------------------------------------------------------------
201/499 validated structures have at least one duplicates
CLEANING
-----------------------------------------------------------------------------
497/499 structures were successfully cleaned
2/499 structure(s) was/were unsuccessfully cleaned and need to be rechecked
-----------------------------------------------------------------------------
No duplicate was found (in 497 cleaned structures)
NORMALIZATION
-----------------------------------------------------------------------------
497/497 structures were successfully normalized
-----------------------------------------------------------------------------
2/497 normalized structures have at least one duplicates
REFINEMENT SUMMARY
-------------------

SMILES      errorTag  \
0   CCC1(O)CC2CN(CCc3c([nH]c4ccccc34)C(C(=O)OC)(c3...       Mixture   
1                                  CCCC(=O)[O-].[Na+]          Salt   
2             COc1ccc(C2CC(=O)c3c([O-])cc(O)cc3O2)cc1        Charge   
3   O=C(O)C1=CC(=CC=[N+]2c3cc(O)c(OC4OC(CO)C(O)C(O...        Charge   
4                                  CCCC(=O)[O-].[Na+]        Charge   
5   CC1CN2CC3CCC(=O)C4=C5C(=C(O)C1CC2C53C)C1=C4CCO...      Tautomer   
6         O=C1NCc2c1c1c3ccccc3[nH]c1c1[nH]c3ccccc3c21      Tautomer   
7   COC(=O)C1=C(CCO)C2=C3C1=C(O)C1(O)CC4N(CC1C)CC(...      Tautomer   
8   CC(C)=CCCC1(C)C(CC=C(C)C)CC2(CC=C(C)C)C(=O)C1(...      Tautomer   
9      CC(C)C1=CC2=CC(=O)C3C(C)(C)CCCC3(C)C2=C(O)C1=O      Tautomer   
10       C=C1CCC2C(C)(CO)C(O)CCC2(C)C1CC=C1C(=O)OCC1O      Tautomer   
11  CC1CCC2(OC1)OC1CC3C4CCC5=CC(=O)CCC5(C)C4CCC3(C...      Tautomer   
12  COC1=C(OC)C(=O)C(C2COc3c(cc(OC)c(OC)c3OC)C2)=C...      Tautomer   
13           C=C1CCCC2C1(C)CCC(C)C2(C)CCC1=CC(=O)OC1O      Tautomer   
14  CC1=CC(=O)C(O)C2(C)C1CC(O)C13COC4(C)C(OC(=O)C(...      Tautomer   
15  CC12CCC3C(CCC4(O)CC(O)CCC34C=O)C1(O)CCC2C1=CC(...      Tautomer   
16  CC(C)=CCc1c(CO)cc2oc3cc(O)c(O)c(CC=C(C)C)c3c(=...      Tautomer   
17  CC1CC(C(C)C2CCC3C4CC5OC56C(O)C=CC(=O)C6(C)C4CC...      Tautomer   
18  CC1=CCCC2C1(C)CCC(C)C2(C)CCC(=CN1C(=O)C=C(CCC2...      Tautomer   
19          C=C1C(=O)OC2C3C(CO)=CC(=O)C3=C(C)CC(O)C12      Tautomer   
20                             O=C1OC(C(O)CO)C(O)=C1O      Tautomer   
21        C=C1C(O)CCC2C1(C)CCC(C)C2(C)CCC1=CC(=O)OC1O      Tautomer   
22  COc1cc(C2c3cc4c(cc3C(=O)C3COC(=O)C32)OCO4)cc(O...      Tautomer   
23  C=C(CCC(C)C1CCC2C3=C(C(=O)C(O)C21C)C1(C)CCC(O)...      Tautomer   
24                CCCCCCCCCCCCCC1=C(O)C(=O)C=C(O)C1=O      Tautomer   
25  CC(=O)CC1C(=O)C(CO)=CCC(C)(CCC=C(C)C)C1C=COC(=...      Tautomer   
26     CCC12C=CCN3CCC4(C(=C(C(=O)OC)C1)Nc1ccccc14)C32      Tautomer   
27   C=C1CCC2C(=C)C(=O)OC2C=C(C)CC2C=C(CCC1=O)C(=O)O2      Tautomer   
28  CC12C(=O)c3cccc(O)c3C(=O)C1CC1=C2C(=O)c2c(O)cc...      Tautomer   
29           CC1=CCCC2C1(C)CCC(C)C2(C)CCC1=CC(=O)OC1O      Tautomer   
30  C=C(C)C1CCC2C3(C)CCC(C(C)C4CC=C(C)C(=O)O4)C3(C...      Tautomer   
31  CC(=O)C1=C(O)C=C2Oc3c(C(C)=O)c(O)c(C)c(O)c3C2(...      Tautomer   
32               COc1cc(O)c2c(c1)OC(c1ccccc1)C(O)C2=O      Tautomer   
33  O=C1c2c(O)cc(OC3OC(CO)C(O)C(O)C3O)cc2OC(c2ccc(...      Tautomer   
34                          CC(C)=CCNc1ncnc2nc[nH]c12      Tautomer   
35       O=C(C=Cc1ccc(O)c(O)c1)OC1CC(C(=O)O)=CC(O)C1O      Tautomer   
36  CC1(C)C2CC(=O)C3(CO2)C2CCC4CC2(C(=O)C(O)C13)C(...      Tautomer   
37  COc1cc2oc3cc(O)c(OC)c(CC=C(C)C)c3c(=O)c2c(O)c1...      Tautomer   
38         O=C1CC=C2C3CC4N(CCC45c4ccccc4N1C25)CC3=CCO      Tautomer   
39              CCCCCCC=CCCCCCCCCCC1(O)C=CC(=O)C(O)C1      Tautomer   
40                   CC1=c2[nH]c3c(c2CCN1)C=CC(=O)C=3      Tautomer   
41  C=C1C(=O)OC2C=C(C)C(=O)C=CC(C)(O)CC(OC(=O)C(C)...      Tautomer   
42  CC(=O)OC1CC(C)C(O)(CCC2=CC(=O)OC2)C2(C)CCCC(C)...      Tautomer   
43  O=C(O)C1=CC(=CC=[N+]2c3cc(O)c(OC4OC(CO)C(O)C(O...      Tautomer   
44      C=CC1(C)CCC2=C(C1)C(=O)CC1C(C)(C(=O)O)CCCC21C      Tautomer   
45         O=c1oc2c(O)c(O)cc3c(=O)oc4c(O)c(O)cc1c4c23      Tautomer   
46  CCCCCCCCCC(O)C1CCC(C2CCC(CCCCCCCCCCC(O)CC3=CC(...      Tautomer   
47  C=C1C(=O)C23CC1CCC2C12COC(CC1=O)C(C)(C)C2C(O)C3=O      Tautomer   
48  CCCCCCCCCCCCC(O)C1CCC(C(O)CCCCC(O)CCCCCC(O)CC2...      Tautomer   
49                COC1=CC=C2C(=O)C3=C(C)C=CNC3=C2C1=O      Tautomer   
50                C=C1C(=O)OC2CC(C)C(C=CC(C)=O)=CCC12      Tautomer   
51  CC(=O)CC1C(=O)C(CO)=CCC(C)(CC=CC(C)(C)O)C1C=CO...      Tautomer   
52  COC(=O)C12OCC34C(CC5C(C)C=C(OC6OC(CO)C(O)C(O)C...      Tautomer   
53  C=C(CCC(C)C1CCC2C3=C(C(=O)CC21C)C1(C)CCC(=O)C(...      Tautomer   
54  COC(=O)C1C2=C(C(=O)C(=O)C(C(C)C)=C2O)C2(C)CCCC...      Tautomer   
55  CC(=O)CC1C(=O)C(CO)=CCC(C)(CCC(O)C(C)(C)O)C1C=...      Tautomer   


**Notice**: `idx` returned by different substage is different from each other. For example, SMILES `idx 277` returned by `rmMixtures()` is different from SMILES `idx 277` returned by `clSalts()` because `rmMixtures()` belongs to substage `validation` while `clSalts()` belongs to substage `cleaning`. When using refineComplete(), the order of substage is `validation`, followed by `cleaning`, and `normalization`.

#### 4.4. Obtaining lists of duplicated refined SMILES(s) (if any)

In [29]:
refineComplete(smiles, removeDuplicates=True, getDuplicatedIdx=True)

VALIDATION
-----------------------------------------------------------------------------
499/500 structures were successfully validated
1/500 structure(s) was/were unsuccessfully validated and need to be rechecked
-----------------------------------------------------------------------------
There are 381 unique structures filtered from 499 initial validated structures
CLEANING
-----------------------------------------------------------------------------
380/381 structures were successfully cleaned
1/381 structure(s) was/were unsuccessfully cleaned and need to be rechecked
-----------------------------------------------------------------------------
No duplicate was found (in 380 cleaned structures)
NORMALIZATION
-----------------------------------------------------------------------------
379/380 structures were successfully normalized
1/380 structure(s) were unsuccessfully normalized and need to be rechecked
-----------------------------------------------------------------------------

idx       matchedIdx         fromFunction
0      0  [0, 9, 69, 278]   validateComplete()
1      3         [3, 467]   validateComplete()
2      7         [7, 215]   validateComplete()
3      9  [0, 9, 69, 278]   validateComplete()
4     11        [11, 418]   validateComplete()
..   ...              ...                  ...
199  495       [213, 495]   validateComplete()
200  496       [472, 496]   validateComplete()
201   na               na      cleanComplete()
202   83         [83, 90]  normalizeComplete()
203   90         [83, 90]  normalizeComplete()

[204 rows x 3 columns]

**Notice**: `idx` returned by different substage is different from each other. For example, SMILES `idx 0` returned by `validateComplete()` is different from SMILES `idx 0` returned by `normalizeComplete()`. For a substage, if `idx na` appears, it means no duplicate was found in the stage. When using `refineComplete()`, the order of substage is `validation`, followed by `cleaning`, and `normalization`.

#### 4.5. Export csv files

In [30]:
# refineOutput= '../refinement/'
# refineComplete(smiles, removeDuplicates=True,  exportCSV=True, outputPath=refineOutput)
# refineComplete(smiles, removeDuplicates=False, exportCSV=True, outputPath=refineOutput)
# refineComplete(smiles, removeDuplicates=True,  getDuplicatedIdx=True,   exportCSV=True, outputPath=refineOutput)
# refineComplete(smiles, removeDuplicates=False, getUnrefinedStruct=True, exportCSV=True, outputPath=refineOutput)

---

### PART 2: STANDARD DATA CURATION

This session introduces a standard data curation procedure proposed by [Fourches et al. (2010)](https://pubs.acs.org/doi/10.1021/ci100176x) whose major substages are represented by functions in substage modules. The data curation process is characterized by the flowchart below.

![Pipeline](https://raw.githubusercontent.com/mldlproject/chemtonic/0fdf3801a315d23cee264535b5bc13bea1b83572/images/pipelinesmall.svg)

This session is designed for you IF
- You've known about this procedure and want to perform data curation process only OR
- You're novice users who are looking for a standard data curation procedure and want to perform to data curation process OR
- You're computational scientists who are looking for a built module to integrate into your to-be-deployed model as a checking gate for input SMILES(s).

#### Completing data curation with one-run task

In [ ]:
refineComplete(smiles)

The function and its default parameters are designed based on [Fourches et al. (2010)](https://pubs.acs.org/doi/10.1021/ci100176x)'s procedure.
```python
refineComplete(compounds, getUnrefinedStruct=False, deSalt=False, neutralize=False, deTautomerize=True, deSterioisomerize=True, removeDuplicates=True, getDuplicatedIdx=False, exportCSV=False, outputPath=None, printlogs=True):

```
To run the task with other conditions, please make sure you understand the role of each substage's steps.

`refineComplete(compounds)` is equal to `normalizeComplete(cleaningComplete(validateComplete(compounds, removeDuplicates=True), removeDuplicates=True), removeDuplicates=True)`.

You can perform each substage to obtain detailed logs as well as extract correct `idx`.

In [ ]:
validated_compounds  = validateComplete(smiles, removeDuplicates=True)
cleaned_compounds    = cleanComplete(validated_compounds, removeDuplicates=True)
normalized_compounds = normalizeComplete(cleaned_compounds, removeDuplicates=True)
refined_compounds    = normalized_compounds

`normalized_compounds` is equal to `refined_compounds` when the input SMILES(s) is/are cleaned.

-END-